up-l1/someip.adoc

```diff
@@ -0,0 +1,265 @@
+
+*SOME/IP* is based on the Internet Protocol (IP) and uses UDP or TCP as the transport layer. It supports various communication patterns, including unicast, multicast, and broadcast. The protocol defines a set of message formats and procedures for encoding and decoding data, handling service discovery, and managing communication sessions.
+By using *SOME/IP*, automotive ECUs can communicate with each other to exchange information related to vehicle functions, such as sensor data, control commands, diagnostics, and software updates. It enables the integration and interoperability of different ECUs from various automotive suppliers within a vehicle's network architecture.
+Overall, *SOME/IP* plays a crucial role in facilitating efficient and standardized communication between ECUs in modern automotive systems.
+
+The following specification will elaborate on how uProtocol shall be mapped to *SOME/IP* protocol so that mechatronics applications and services can be connected to non-mechatronics and cloud based applications and services.
+
+=== Open Source Libraries
+
+As of the time of writing, there are two open source projects that are available for *SOME/IP*:
+
+1. *https://github.com/COVESA/vsomeip[COVESA-vsomeip]:* C++ solution based on the original specifications developed by BMW prior to contribution to AUTOSAR
+2. *https://projects.eclipse.org/projects/automotive.sommr[Eclipse-SommR]:*  Rust implementation based on the latest AUTOSAR *SOME/IP* specifications.
+
+NOTE: Given the latest version of *SOME/IP* specifications from AUTOSAR is not open source, we cannot share or implement uProtocol using Eclipse-SommR.
+
+=== AUTOSAR References
+All references below to:
+
+- `[PRS_SOMEIP_XXXXX]` refer to link:https://some-ip.com/standards.shtml[AUTOSAR *SOME/IP* specification].
+- `[PRS_SOMEIP_SD_XXXXX]` refer to link:https://some-ip.com/standards.shtml[SOME/IP Service Discovery Protocol Specification].
+
+A search engine usually finds the correct document based on the reference number.
+
+== Transport Layer Translation
+
+=== UAttributes
+
+The following table highlights the mapping of *uAttributes* to *SOME/IP* fields that apply to all event types. For more information about *uAttributes*, please refer to link:../basics/uattributes.adoc[uAttributes Specification].
+
+
+==== Uris
+
+link:../basics/uattributes.adoc[uAttributes Specification] explains that `source` attributes defines the address of whom sent the message, while `sink` defines the destination for the message in link:../basics/uri.adoc[`UUri`] format. `UUri` is translated to *SOME/IP* attributes per <<sink-source-uri-mapping>> below.
+
+.Sink & Source URI Mapping
+[sink-source-uri-mapping, cols="1,1,2"]
+|===
+| *UUri Portion* | *SOME/IP Field* | *Description*
+a| `authority_name` | IP address & port
+a| IP address & port (destination endpoint) of the mDevice.
```

Comment:
I think we need to elaborate more on this portion. The URI can be either a host or IP address. If the authority_name is host, we need to then resolve to an IP address in the transport implementation.

Reply:
Curious to learn more here too.

Reply:
`vsomeip` gets the required endpoint information (via service discovery or static json configuration), but they have decided to leave it private for the implementation, so you have a better abstraction of SOME/IP services/instances/methods available in the network.

e.g. it is possible to have multiple endpoints (hosts) offering the same service, but with different instance or version. vsomeip router aggregates all discovered service/instance information and maps the endpoint to service/instance/major_ver/minor_ver/[method | eventgrup/event]
API requires service and instance as minimum.

The only way I see to provide the mDevice IP:port in `authority_name` is to have it pre-configured (mapped to service/instance pair).

If "host" means any string, it could be created from service/instance pair, e.g. "<service>.<instance>v<version>", but then that same information is present also in ue_id, ue_version_major